src/Storages/MergeTree/MergeTreeBlockOutputStream.cpp

```diff
@@ -17,18 +23,43 @@ void MergeTreeBlockOutputStream::write(const Block & block)
     storage.delayInsertOrThrowIfNeeded();
 
     auto part_blocks = storage.writer.splitBlockIntoParts(block, max_parts_per_block, metadata_snapshot);
+    MergeTreeData::DataPartsVector inserted_parts;
     for (auto & current_block : part_blocks)
     {
         Stopwatch watch;
 
         MergeTreeData::MutableDataPartPtr part = storage.writer.writeTempPart(current_block, metadata_snapshot);
         storage.renameTempPartAndAdd(part, &storage.increment);
+        inserted_parts.push_back(part);
 
         PartLog::addNewPart(storage.global_context, part, watch.elapsed());
 
-        /// Initiate async merge - it will be done if it's good time for merge and if there are space in 'background_pool'.
-        if (storage.merging_mutating_task_handle)
+        if (auto part_in_memory = asInMemoryPart(part))
+        {
+            storage.in_memory_merges_throttler.add(part_in_memory->block.bytes(), part_in_memory->rows_count);
+
+            if (storage.merging_mutating_task_handle && !storage.in_memory_merges_throttler.needDelayMerge())
+            {
+                storage.in_memory_merges_throttler.reset();
+                storage.merging_mutating_task_handle->signalReadyToRun();
+            }
+        }
+        else if (storage.merging_mutating_task_handle)
+        {
+            /// Initiate async merge - it will be done if it's good time for merge and if there are space in 'background_pool'.
             storage.merging_mutating_task_handle->signalReadyToRun();
+        }
+    }
+
+    if (storage.getSettings()->in_memory_parts_insert_sync)
+    {
+        for (const auto & part : inserted_parts)
+        {
+            auto part_in_memory = asInMemoryPart(part);
+            if (!part_in_memory->waitUntilMerged(in_memory_parts_timeout))
```

Comment:
And what user should do with this error? I think most of the users will just retry this error and duplicate data.

Reply:
I've removed this functionality at all. It's better to implement sync insert, that depends on fsync. I think this guaranty will be enough.